### PR TITLE
Split Onnxruntime Nuget GPU package

### PR DESCRIPTION
**Pull request by mszhanyi**
### Description
1. Update donwload-artifacts to flex-downloadartifacts to make it eaiser to debug.
2. Move the native files into Gpu.Windows and Gpu-linux packages. Onnxruntime-Gpu has dependency on them.
3. update the package validation as well
4. Add 2 stages to run E2E test for GPU.Windows and GPU.Linux
   for example:
   
image

### Motivation and Context
Single Onnxruntime.Gpu Package size has already excceded the Nuget size limit.
We split the package into some smaller packages to make them can be published.

For compatibility, the user can install or upgrade Onnxruntime.Gpu, which will install Gpu.Windows and Gpu.Linux automatically.
And the user can only install Gpu.Windows and Gpu.Linux directly. 

### Test Link
1. In ORT_NIGHTLY

2.  Install the preview version in nuget-int.  (nuget source: https://apiint.nugettest.org/v3/index.json)

### CODE REVIEW

**Comment by skottmckay** [comment]
>         )

> "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."

Can we adjust this for the platform specific packages as they will not have artifacts for all platforms?

---
Refers to: tools/nuget/generate_nuspec_for_native_nuget.py:178 in 79d3b00. [](commit_id = 79d3b00cb47d63861d14e28eb8985a45059159bf, deletion_comment = False)

**Comment by mszhanyi** [comment]
> > ```
> >     )
> > ```
> 
> > "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."
> 
> Can we adjust this for the platform specific packages as they will not have artifacts for all platforms?
> 
> Refers to: tools/nuget/generate_nuspec_for_native_nuget.py:178 in [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf). [](commit_id = [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf), deletion_comment = False)

Thank your suggestion. I think it needs to make a further discussion with PM. Because it will change current User experience, that is, the package name must be change and the old package might not be upgraded directly.

File: tools/nuget/validate_package.py [hunk 1]
```diff
@@ -117,24 +118,33 @@ def check_if_dlls_are_present(
...
                 for dll in win_gpu_package_libraries:
                     path = folder + "/" + dll
                     print("Checking path: " + path)
                     if path not in file_list_in_package:
                         print(dll + " not found for " + platform)
                         raise Exception(dll + " not found for " + platform)
+            # In Nuget GPU package, gpu header files are not in dependent package.
+            if (is_gpu_package and package_type != "nuget") or (
+                package_type == "nuget" and not is_gpu_dependent_package
+            ):
```
Comment (skottmckay) [inline tools/nuget/validate_package.py:147]: This logic is a little hard to follow. Would it be simpler if we set a couple of variables and check for the library/headers based on that? 

e.g. 
```py
package_contains_library = False if package_type == "nuget" and is_gpu_package else True
package_contains_headers = False if is_gpu_dependent_package else True
```

Would be good to check the package does not contain headers as well when that is the expected state.

File: tools/nuget/validate_package.py [hunk 2]
```diff
@@ -276,9 +294,11 @@ def validate_nuget(args):
     full_nuget_path = os.path.join(args.package_path, nuget_file_name)
 
     if "Gpu" in nuget_file_name:
-        is_gpu_package = True
+        is_gpu_package = bool("Gpu-" not in nuget_file_name)
+        is_gpu_dependent_package = bool("Gpu-" in nuget_file_name)
```
Comment (skottmckay) [inline tools/nuget/validate_package.py:298]:
```suggestion
        is_gpu_package = nuget_file_name.endswith('Gpu')
        is_gpu_dependent_package = not is_gpu_package
```

**Comment by skottmckay** [comment]
> > > "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."
> > 
> > 
> > Can we adjust this for the platform specific packages as they will not have artifacts for all platforms?
> > Refers to: tools/nuget/generate_nuspec_for_native_nuget.py:178 in [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf). [](commit_id = [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf), deletion_comment = False)
> 
> Thank your suggestion. I think it needs to make a further discussion with PM. Because it will change current User experience, that is, the package name must be changed and the old package might not be upgraded directly.

I'm only suggesting updating the text that is in the package description not the package names. It's misleading to read a description that doesn't match the package contents.

Ideally the linux/windows gpu package descriptions will say it contains the native shared library for linux/windows, and the base GPU package could maybe say it contains the native artifacts via the other two nuget package names.

File: csharp/OnnxRuntime.CSharp.proj [hunk 3]
```diff
@@ -20,6 +20,7 @@ CMake creates a target to this project
     <IsWindowsBuild Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::Windows)))' == 'true'">true</IsWindowsBuild>
     <IsLinuxBuild Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::Linux)))' == 'true'">true</IsLinuxBuild>
     <ExecutionProvider Condition=" '$(ExecutionProvider)' == '' ">None</ExecutionProvider>
+    <IsDependentPackage Condition=" '$(IsDependentPackage)' == '' ">false</IsDependentPackage>
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply csharp/OnnxRuntime.CSharp.proj:23]: or maybe 'IsPlatformSpecificSubPackage'?

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 4]
```diff
@@ -797,13 +810,45 @@ stages:
...
         platform: 'Any CPU'
         msbuildArguments: '-t:CreatePackage -p:OnnxRuntimeBuildDirectory="$(Build.BinariesDirectory)" -p:OrtPackageId=Microsoft.ML.OnnxRuntime.Gpu -p:IsReleaseBuild=${{ parameters.IsReleaseBuild }} -p:ReleaseVersionSuffix=$(ReleaseVersionSuffix)'
         workingDirectory: '$(Build.SourcesDirectory)\csharp'
+      env:
+        # check out csharp/OnnxRuntime.CSharp.proj for the usage of these variables
+        CurrentDate: $(CurrentDate)
+        CurrentTime: $(CurrentTime)
+
+    - task: MSBuild@1
+      displayName: 'Build Nuget Packages Microsoft.ML.OnnxRuntime.Gpu-win'
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:826]: Thanks. And I would like to rename it to Gpu.sub.linux/Gpu.sub.windows. It could reminder users not to install it alone.

File: tools/nuget/validate_package.py (hunk 2 above)
Comment (mszhanyi) [reply tools/nuget/validate_package.py:298]: We couldn't use endswith because the package name includes the version string.

File: csharp/OnnxRuntime.CSharp.proj (hunk 3 above)
[the remark this answers is not in the document]
Comment (mszhanyi) [reply csharp/OnnxRuntime.CSharp.proj]: thx, updated

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 5]
```diff
@@ -324,6 +347,8 @@ def generate_files(line_list, args):
     ]
     is_mklml_package = args.package_name == "Microsoft.ML.OnnxRuntime.MKLML"
     is_cuda_gpu_package = args.package_name == "Microsoft.ML.OnnxRuntime.Gpu"
+    is_cuda_gpu_win_package = args.package_name == "Microsoft.ML.OnnxRuntime.Gpu-win"
+    is_cuda_gpu_linux_package = args.package_name == "Microsoft.ML.OnnxRuntime.Gpu-linux"
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: thx, updated

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 6]
```diff
@@ -39,15 +39,18 @@ def get_package_name(os, cpu_arch, ep, is_training_package):
 # Currently we take onnxruntime_providers_cuda from CUDA build
 # And onnxruntime, onnxruntime_providers_shared and
 # onnxruntime_providers_tensorrt from tensorrt build
-def is_this_file_needed(ep, filename):
+# cuda binaries are in dependent packages not in Microst.ML.OnnxRuntime.Gpu
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: thx

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 7]
```diff
@@ -389,23 +414,31 @@ def generate_files(line_list, args):
     runtimes = f'{runtimes_target}{args.target_architecture}\\{runtimes_native_folder}"'
 
     # Process headers
+    build_dir = "buildTransitive" if "Microsoft.ML.OnnxRuntime.Gpu" in args.package_name else "build"
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: Thanks reminder.
No, it doesn't work with build.py now. 
Because I didn't want to make OnnxRuntime.CSharp.proj too complex.
I thought it'd better that it generates one nuget package (except onnxruntime.managed) for OnnxRuntime.CSharp.proj.

[the remark this answers is not in the document]
Comment (skottmckay) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: We do have folks that use build.py with `--build_nuget` to test. 

Do we maybe need a setting saying whether to split up the GPU package or not so that for local testing it builds a single package?

[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: I've moved sub packages generation into csproj file, thus, build.py and CI both work.

File: tools/nuget/validate_package.py (hunk 1 above)
Comment (mszhanyi) [reply tools/nuget/validate_package.py:147]: thx. updated

File: tools/nuget/generate_nuspec_for_native_nuget.py
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py:1]: Good finding!
I think it was caused by https://github.com/microsoft/onnxruntime/pull/18819#discussion_r1429844633.

[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py:1]: I double checked that both netstandard2.0/2.1 have same props and targets file

File: csharp/OnnxRuntime.CSharp.proj [hunk 8]
```diff
@@ -144,6 +147,87 @@ CMake creates a target to this project
     />
   </Target>
 
+  <!-- To support build.py and CI easily, generate the sub packages in the proj -->
+  <Target Name="CreateSubGpuWinPackages" AfterTargets="CreatePackage" Condition="'$(OrtPackageId)' == 'Microsoft.ML.OnnxRuntime.Gpu'">
+    <!-- Manually create the nuspec for the native Microsoft.ML.OnnxRuntime package -->
+    <Message Importance="High" Text="Generating nuspec for the native Microsoft.ML.OnnxRuntime nuget package..." />
+    <Exec Command="$(PythonExe) $(GenerateNuspecScript) --package_version $(PackageVersion) --package_name Microsoft.ML.OnnxRuntime.Sub.Gpu.Windows --target_architecture $(TargetArchitecture) --build_config $(Configuration) --native_build_path $(NativeBuildOutputDirAbs) --packages_path $(OnnxRuntimePackagesDirectoryAbs) --ort_build_path $(OnnxRuntimeBuildDirectoryAbs) --sources_path $(OnnxRuntimeSourceDirectoryAbs) --commit_id $(GitCommitHash) --is_release_build $(IsReleaseBuild) --execution_provider $(ExecutionProvider)"
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:154]: Comment/message needs updating here and line 192 - 194.

I find the package name a little confusing due to the ordering. Having 'OnnxRuntime.Sub.Gpu' makes me think it's related to the base OnnxRuntime package. Microsoft.ML.OnnxRuntime.Gpu.Sub.Windows would have a clearer association with Microsoft.ML.OnnxRuntime.Gpu.

Having thought about this a little more and aking a step back, would it be simpler if we instead made both the platform specific packages standalone (include headers, props and targets files in both) and call them Microsoft.ML.OnnxRuntime.Gpu.Windows and Microsoft.ML.OnnxRuntime.Gpu.Linux? We would still have Microsoft.ML.OnnxRuntime.Gpu with a dependency on both for backwards compatibility and ease of use, but it would only provide those dependencies. That setup avoids having to try and explain what a 'sub' package is to confused users.

File: csharp/OnnxRuntime.CSharp.proj [hunk 9]
```diff
@@ -144,6 +147,87 @@ CMake creates a target to this project
...
+          ContinueOnError="False"
+          ConsoleToMSBuild="true">
+        <Output TaskParameter="ConsoleOutput" PropertyName="GenerateNuspecOutput" />
+    </Exec>
+    <!-- run `nuget pack` on Windows or `dotnet pack` on Linux to create the native nupkg -->
+    <Message Importance="High" Text="Bundling native shared library artifacts into nuget package..." />
+    <Exec Condition=" '$(IsWindowsBuild)' == 'true'"
+          Command="$(NugetExe) pack NativeNuget.nuspec"
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:162]: If we overwrite NativeNuget.nuspec for each package it will be harder to debug issues. Can we have a different nuspec filename for each package to avoid that?

File: csharp/OnnxRuntime.CSharp.proj [hunk 10]
```diff
@@ -144,6 +147,87 @@ CMake creates a target to this project
...
+          ContinueOnError="False"
+          ConsoleToMSBuild="true">
+      <Output TaskParameter="ConsoleOutput" PropertyName="OutputOfExec" />
+    </Exec>
+
+    <!-- build.py uses dotnet to build on linux so we know it's available.
+         nuget needs to be run using mono to work correctly, but installing mono on WSL breaks interop
+         (see https://github.com/microsoft/WSL/issues/8531). in order to play nicely with both we use `dotnet pack`
+         to pack the native nuget package using a stub csproj to provide the nuspec file path.
+    -->
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:173]: nit: is having this message 3 times of value, or is it enough just to have in the original code?

File: csharp/OnnxRuntime.CSharp.proj [hunk 11]
```diff
@@ -144,6 +147,87 @@ CMake creates a target to this project
     />
   </Target>
 
+  <!-- To support build.py and CI easily, generate the sub packages in the proj -->
+  <Target Name="CreateSubGpuWinPackages" AfterTargets="CreatePackage" Condition="'$(OrtPackageId)' == 'Microsoft.ML.OnnxRuntime.Gpu'">
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:151]: Instead of cut-and-pasting can we define properties for the values that change each time (may just be package name) and use CallTarget?

e.g. something like this: https://stackoverflow.com/a/2017647/684911

**Comment by skottmckay** [comment]
> > > > "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."
> > > 
> > > 
> > > Can we adjust this for the platform specific packages as they will not have artifacts for all platforms?
> > > Refers to: tools/nuget/generate_nuspec_for_native_nuget.py:178 in [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf). [](commit_id = [79d3b00](https://github.com/microsoft/onnxruntime/commit/79d3b00cb47d63861d14e28eb8985a45059159bf), deletion_comment = False)
> > 
> > 
> > Thank your suggestion. I think it needs to make a further discussion with PM. Because it will change current User experience, that is, the package name must be changed and the old package might not be upgraded directly.
> 
> I'm only suggesting updating the text that is in the package description not the package names. It's misleading to read a description that doesn't match the package contents.
> 
> Ideally the linux/windows gpu package descriptions will say it contains the native shared library for linux/windows, and the base GPU package could maybe say it contains the native artifacts via the other two nuget package names.

I'd still prefer to see this updated based on the package name.

e.g. add 
```py
    elif "Microsoft.ML.OnnxRuntime.Gpu.Sub.Linux" in package_name:  
        description = (
            "This package contains Linux native shared library artifacts for ONNX Runtime with CUDA."
        )
    elif "Microsoft.ML.OnnxRuntime.Gpu.Sub.Windows" in package_name:  
        description = (
            "This package contains Windows native shared library artifacts for ONNX Runtime with CUDA."
        )
    elif "Microsoft.ML.OnnxRuntime" in package_name:  # This is a Microsoft.ML.OnnxRuntime.* package
        description = (
            "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."
        )
```

File: csharp/OnnxRuntime.CSharp.proj (hunk 8 above)
Comment (mszhanyi) [reply csharp/OnnxRuntime.CSharp.proj:154]: Do you mean that we should support Microsoft.ML.Onnxruntime.Gpu.Windows be installed alone?

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 12]
```diff
@@ -57,7 +60,11 @@ def generate_file_list_for_ep(nuget_artifacts_dir, ep, files_list, include_pdbs,
                 child = child / "lib"  # noqa: PLW2901
                 for child_file in child.iterdir():
                     suffixes = [".dll", ".lib", ".pdb"] if include_pdbs else [".dll", ".lib"]
-                    if child_file.suffix in suffixes and is_this_file_needed(ep, child_file.name):
+                    if (
+                        child_file.suffix in suffixes
+                        and is_this_file_needed(ep, child_file.name, package_name)
+                        and package_name != "Microsoft.ML.OnnxRuntime.Gpu-linux"
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: It'd better to keep it as before.
Because we only split win/linux files for gpu packages.
It's clear that there's special changes for GPU package.
I would like to be defensive to avoid regression.

[the remark this answers is not in the document]
Comment (skottmckay) [reply tools/nuget/generate_nuspec_for_native_nuget.py]: Can we maybe move `is_this_file_needed` to define it inside of `generate_file_list_for_ep`? It has a very generic name but it's usage is specific to that function. Doing that will make it easier to see the multiple places where there is special logic for the GPU package.

File: csharp/OnnxRuntime.CSharp.proj (hunk 11 above)
Comment (mszhanyi) [reply csharp/OnnxRuntime.CSharp.proj]: Thank your suggestion. updated

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 13]
```diff
@@ -161,11 +176,18 @@ def generate_description(line_list, package_name):
...
             + "more details."
         )
+    elif "Microsoft.ML.OnnxRuntime.Gpu.Linux" in package_name:
+        description = ("This package contains Linux native shared library artifacts for ONNX Runtime with CUDA.")
+    elif "Microsoft.ML.OnnxRuntime.Gpu.Windows" in package_name:
+        description = ("This package contains Windows native shared library artifacts for ONNX Runtime with CUDA.")
+    elif "Microsoft.ML.OnnxRuntime.Gpu" in package_name:
+        description = (
+            "This package contains native shared library artifacts for all supported platforms of ONNX Runtime."
+        )
```
Comment (skottmckay) [inline tools/nuget/generate_nuspec_for_native_nuget.py:186]:
```suggestion
```

This is handled by the next `elif`

File: csharp/OnnxRuntime.CSharp.proj (hunk 8 above)
Comment (mszhanyi) [reply csharp/OnnxRuntime.CSharp.proj]: Now, the Onnxruntime.GPU.Windows/Linux can work indivisually.

File: csharp/OnnxRuntime.CSharp.proj [hunk 14]
```diff
@@ -91,23 +98,10 @@ CMake creates a target to this project
...
-
-    <!-- Create Microsoft.ML.OnnxRuntime.Managed with the C# bindings using the C# project -->
-    <Message Importance="High" Text="Creating Microsoft.ML.OnnxRuntime.Managed nuget package..." />
-    <MSBuild Projects="src\Microsoft.ML.OnnxRuntime\Microsoft.ML.OnnxRuntime.csproj"
-             Targets="CopyMiscFiles;Pack"
-            Properties="NoBuild=true;Platform=AnyCPU;PackageVersion=$(PackageVersion);OrtPackageId=$(OrtPackageId);IncludeMobileTargets=$(IncludeMobileTargets)"/>
-
+  <Target Name="CreateNativePackage">
     <!-- Manually create the nuspec for the native Microsoft.ML.OnnxRuntime package -->
     <Message Importance="High" Text="Generating nuspec for the native Microsoft.ML.OnnxRuntime nuget package..." />
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:103]: nit: optional as it's an existing issue I just noticed, but it would be nice if we used `$(OrtPackageId)` instead of the hardcoded `Microsoft.ML.OnnxRuntime` so the actual package name is output in the message.

Comment (mszhanyi) [reply csharp/OnnxRuntime.CSharp.proj]: thx. updated

**Review by skottmckay — DISMISSED** [review]

**Review by skottmckay — APPROVED** [review]
